lib/libc/gen/fdopendir.c

```diff
@@ -49,8 +49,14 @@
 DIR *
 fdopendir(int fd)
 {
+	int flags;
 
-	if (_fcntl(fd, F_SETFD, FD_CLOEXEC) == -1)
+	flags = _fcntl(fd, F_GETFD, 0);
+	if (flags == -1)
+		return (NULL);
+
+	if ((flags & FD_CLOEXEC) == 0 &&
+	    _fcntl(fd, F_SETFD, flags | FD_CLOEXEC) == -1)
```

Comment:
I do not like this, it is racy.  Suppose that another thread manipulates the same fd flag in parallel.  Then our flags is stale and we might still clear a flag set by the other thread.

I suspect we need a new fcntl op to just set the passed bits of flags, like F_ORFD or such.

Reply:
> I do not like this, it is racy. Suppose that another thread manipulates the same fd flag in parallel. Then our flags is stale and we might still clear a flag set by the other thread.
> 
> I suspect we need a new fcntl op to just set the passed bits of flags, like F_ORFD or such.

Can it be done in another PR?  The scope is already wide for this one.

Reply:
Definitely, and I think that existing merged PR should not be continued.

But this patch for libc should not go as is, F_ORFD (and might be F_XORFD to clear) needs to be used there.